### PR TITLE
(BOLT-360) Add facts to targets

### PR DESCRIPTION
**Pull request by lucywyman**

### CODE REVIEW

File: bolt-modules/boltlib/lib/puppet/functions/add_facts.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  def add_facts(target, facts)
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue.from_issue_and_stack(
+        Puppet::Pops::Issues::TASK_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, operation: 'add_facts'
+      )
+    end
+
+    inventory = Puppet.lookup(:bolt_inventory) { nil }
+
+    inventory.add_facts(target, facts)
```
Comment (MikaelSmith) [inline bolt-modules/boltlib/lib/puppet/functions/add_facts.rb]: We should test for `nil` inventory before using it. I think other functions raise an error then.

File: bolt-modules/boltlib/lib/puppet/functions/facts.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  def facts(target)
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue.from_issue_and_stack(
+        Puppet::Pops::Issues::TASK_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, operation: 'facts'
+      )
+    end
+
+    inventory = Puppet.lookup(:bolt_inventory) { nil }
+
+    inventory.facts(target)
```
Comment (MikaelSmith) [inline bolt-modules/boltlib/lib/puppet/functions/facts.rb]: Also check for `nil` inventory here.

File: lib/bolt/inventory.rb [hunk 3]
```diff
@@ -108,6 +109,15 @@ def vars(target)
       @target_vars[target.name]
     end
 
+    def add_facts(target, new_facts = {})
+      # TODO: Should this warn if facts is empty?
```
Comment (MikaelSmith) [inline lib/bolt/inventory.rb]: If `new_facts` is empty? I'm on the fence. I can imagine scenarios where you might update facts, but expect that some are empty and not want to have to separately filter them. But also that empty facts could be a warning of something going wrong.

I guess let's warn for now.

File: lib/bolt/inventory.rb [hunk 4]
```diff
@@ -193,5 +201,13 @@ def set_vars_from_hash(target_name, data)
       end
     end
     private :set_vars_from_hash
+
+    def set_facts(tname, hash)
+      if hash
+        @target_facts[tname] = @target_facts[tname] || {}
```
Comment (MikaelSmith) [inline lib/bolt/inventory.rb]: Can be `@target_facts[tname] ||= {}`.

File: lib/bolt/inventory.rb [hunk 5]
```diff
@@ -126,12 +136,10 @@ def update_target(target)
...
 
-      unless data['vars']
-        @logger.debug("Did not find any variables for #{target.name} in inventory")
-        data['vars'] = {}
-      end
-
-      set_vars_from_hash(target.name, data['vars'])
+      # These should only get set from the inventory if they have not yet
+      # been instantiated
+      set_vars_from_hash(target.name, data['vars']) unless @target_vars[target.name]
```
Comment (MikaelSmith) [inline lib/bolt/inventory.rb]: So this initializes them the 1st time from the inventory data, then all subsequent lookups will use whatever latest is. 👍

File: spec/integration/inventory_spec.rb [hunk 6]
```diff
@@ -88,6 +91,21 @@
         expect(run_cli_json(var_plan)[0]['result']['stdout']).to eq(output)
       end
     end
+
+    context 'with facts set' do
+      let(:fact_plan) { ['plan', 'run', 'facts', "host=#{target}"] + config_flags }
+      # TODO: ensure deep merge works
```
Comment (MikaelSmith) [inline spec/integration/inventory_spec.rb]: TODO!

File: lib/bolt/inventory.rb (hunk 5 above)
Comment (lucywyman) [reply lib/bolt/inventory.rb]: Yep :P

Comment (MikaelSmith) [reply lib/bolt/inventory.rb]: Is there a test that covers this particular fix?

Comment (MikaelSmith) [reply lib/bolt/inventory.rb]: Ah, I see it in spec/pal/vars_spec.rb.

File: lib/bolt/inventory.rb [hunk 7]
```diff
@@ -108,6 +109,15 @@ def vars(target)
       @target_vars[target.name]
     end
 
+    def add_facts(target, new_facts = {})
+      @logger.warn("No facts to add")
```
Comment (MikaelSmith) [inline lib/bolt/inventory.rb]: This warning should only happen if `new_facts` is empty.

Comment (lucywyman) [reply lib/bolt/inventory.rb]: Ack, good catch!

**Comment by MikaelSmith** [comment]
Seems this didn't need a rebase. Going to try to get CI working first though.

**Pull request merged**